Sources/StreamChat/StateLayer/MessageSearch.swift

```diff
@@ -0,0 +1,113 @@
+    
+    init(client: ChatClient, environment: Environment = .init()) {
+        authenticationRepository = client.authenticationRepository
+        messageUpdater = environment.messageUpdaterBuilder(
+            client.config.isLocalStorageEnabled,
+            client.messageRepository,
+            client.databaseContainer,
+            client.apiClient
+        )
+        state = environment.stateBuilder(client.databaseContainer)
+    }
+    
+    /// An observable object representing the current state of the search.
+    public let state: MessageSearchState
+    
+    /// Searches for messages with the specified full text search text and updates ``MessageSearchState.messages``.
+    ///
+    /// - Parameter text: A string to search for (which is a full text search).
+    ///
+    /// - Throws: An error while communicating with the Stream API.
+    /// - Returns: An array of paginated chat messages matching to the search term.
+    @discardableResult public func search(text: String) async throws -> [ChatMessage] {
+        // Clear results when there is no text
+        if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, let query = state.query {
+            try await messageUpdater.clearSearchResults(for: query)
+        }
+        
+        let currentUserId = try currentUserId()
+        let query = MessageSearchQuery(
+            channelFilter: .containMembers(userIds: [currentUserId]),
+            messageFilter: .autocomplete(.text, text: text),
+            sort: [.init(key: .createdAt, isAscending: false)]
+        )
+        return try await search(query: query)
+    }
+    
+    /// Searches for messages with the specified query and updates ``MessageSearchState.messages``.
+    ///
+    /// - Parameter query: The search query specifying filters, sorting and pagination.
+    ///
+    /// - Throws: An error while communicating with the Stream API.
+    /// - Returns: An array of paginated chat messages matching to the query.
+    @discardableResult public func search(query: MessageSearchQuery) async throws -> [ChatMessage] {
+        var query = query
+        query.filterHash = state.explicitFilterHash
+        let result = try await messageUpdater.search(query: query, policy: .replace)
+        await state.set(query: query, cursor: result.payload.next)
+        return result.models
+    }
+    
+    /// Searches for more messages matching with the last search query.
+    ///
+    /// - Parameter limit: The limit for the page size. The default limit is 25.
+    ///
+    /// - Throws: An error while communicating with the Stream API.
+    /// - Returns: A next page of chat messages matching to the last query.
+    @discardableResult public func loadNextMessages(limit: Int? = nil) async throws -> [ChatMessage] {
+        guard let query = state.query else { throw ClientError("Call search() before calling for next page") }
+        let limit = (limit ?? query.pagination?.pageSize) ?? Int.messagesPageSize
```

Comment:
Probably out of scope, but ideally we want to remove these `Int.messagesPageSize` public extensions for v5. We should provide a config in `ChatClientConfig` with the default messages pagination size.

Reply:
Added a TODO item since I have used these in many other places as well. Needs to be changed with one go.

Reply:
good point, but yes, out of scope for this one.